### PR TITLE
Incremental fix for #35 - MERGEABLE CONDITIONNALY

### PR DESCRIPTION
**Pull request by jfalcou**
This PR is failing on purpose up to the point everything will be set as per #35 steps

### CODE REVIEW

File: include/eve/arch/cpu/spec.hpp [hunk 1]
```diff
@@ -22,9 +22,23 @@
...
+  inline constexpr bool supports_avx512popcntdq_    = spy::supports::avx512::popcntdq_;
+  inline constexpr bool supports_avx512_4fmaps_     = spy::supports::avx512::_4fmaps_;
+  inline constexpr bool supports_avx512vnniw_       = spy::supports::avx512::vnniw_;
+  inline constexpr bool supports_avx512vbmi_        = spy::supports::avx512::vbmi_;
+  inline constexpr bool supports_avx512bf16_        = spy::supports::avx512::bf16_;
+  inline constexpr bool supports_avx512bitalg_      = spy::supports::avx512::bitalg_;
+  inline constexpr bool supports_avx512vbmi2_       = spy::supports::avx512::vbmi2_;
+  inline constexpr bool supports_avx512vnni_        = spy::supports::avx512::vnni_;
+  inline constexpr bool supports_avx512vpintersect_ = spy::supports::avx512::vpintersect_;
+}
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/spec.hpp:44]: new lines :)
You gotta check your editor set up

File: include/eve/arch/x86/tags.hpp [hunk 2]
```diff
@@ -17,6 +17,15 @@
...
 namespace eve
 {
+  namespace detail
+  {
+# if defined(SPY_SIMD_IS_X86_AVX512)
+    static inline constexpr bool x86_logical_status = false;
+# else
+    static inline constexpr bool x86_logical_status = true;
+# endif
+  }
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/x86/tags.hpp:27]: What's this?

File: include/eve/arch/cpu/spec.hpp [hunk 3]
```diff
@@ -22,9 +22,23 @@
...
+  inline constexpr bool supports_avx512vl           = spy::supports::avx512::vl_;
+  inline constexpr bool supports_avx512popcntdq_    = spy::supports::avx512::popcntdq_;
+  inline constexpr bool supports_avx512_4fmaps_     = spy::supports::avx512::_4fmaps_;
+  inline constexpr bool supports_avx512vnniw_       = spy::supports::avx512::vnniw_;
+  inline constexpr bool supports_avx512vbmi_        = spy::supports::avx512::vbmi_;
+  inline constexpr bool supports_avx512bf16_        = spy::supports::avx512::bf16_;
+  inline constexpr bool supports_avx512bitalg_      = spy::supports::avx512::bitalg_;
+  inline constexpr bool supports_avx512vbmi2_       = spy::supports::avx512::vbmi2_;
+  inline constexpr bool supports_avx512vnni_        = spy::supports::avx512::vnni_;
+  inline constexpr bool supports_avx512vpintersect_ = spy::supports::avx512::vpintersect_;
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/spec.hpp:43]: Ooooffff - this will get asm testing more unpleasant :)

File: include/eve/arch/x86/tags.hpp [hunk 4]
```diff
@@ -42,8 +52,28 @@ namespace eve
...
+    static constexpr std::size_t expected_cardinal = bytes / sizeof(Type);
+  };
+
+  struct avx2_    : avx_ {};
+
+  struct avx512_  : avx2_
+  {
+    static constexpr std::size_t bits               = 512;
+    static constexpr std::size_t bytes              = 64;
+    static constexpr bool        is_bitwise_logical = detail::x86_logical_status;
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/x86/tags.hpp:72]: ? I see one struct with `is_bit_logical` the other with `is_bitwise_logical`.
Not that either of them makes sense to me

File: include/eve/arch/x86/tags.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/x86/tags.hpp:27]: This also seems a bit off.

`logical<wide<T, N, x86_256>>`

should not be bits. Or should it be bits anyways? 

Or should it also be ok to have

`logical<wide<char, eve::fixed<32>, _x86_512>`?

File: include/eve/arch/cpu/spec.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/arch/cpu/spec.hpp:44]: Dang XD

File: include/eve/arch/x86/tags.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/arch/x86/tags.hpp:27]: When you activate AVX512, all logical values including the one from SSE/AVX have to be represented using the mmask registers. So I globally detect if AVX512 is up and fix this flag to true/false for later usage. It's a refactoring of not wanting the macro everywhere down

File: include/eve/arch/cpu/spec.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/arch/cpu/spec.hpp:43]: According to https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512 , they are usually grouped by blocks of micro architectures. SO Testing for "skylake-like" or "cannonlake-like" is maybe a better idea than testing all of them separately.

File: include/eve/arch/x86/tags.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/arch/x86/tags.hpp:72]: Obvious error is obvious.
The name is also maybe to rework. It's the indicator that the logical is either a regular register OR a compact bitmap register like mmask.

Comment (jfalcou) [reply include/eve/arch/x86/tags.hpp:72]: `logical_use_regular_register` is maybe better

Comment (jfalcou) [reply include/eve/arch/x86/tags.hpp:72]: This one structure was redundant

**Pull request merged**